src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1252,5 +1245,236 @@ internal bool CanAddParsedValue(HttpHeaderParser parser)
+        }
+
+        private ref object? GetValueRefOrAddDefault(HeaderDescriptor key)
+        {
+            object? store = _headerStore;
+            if (store is HeaderEntry[] entries)
+            {
+                for (int i = 0; i < _count && i < entries.Length; i++)
+                {
+                    if (key.Equals(entries[i].Key))
+                    {
+                        return ref entries[i].Value!;
+                    }
+                }
+
+                int count = _count;
+                _count++;
+                if ((uint)count < (uint)entries.Length)
+                {
+                    entries[count].Key = key;
+                    return ref entries[count].Value!;
+                }
+
+                return ref GrowEntriesAndAddDefault(key);
+            }
+            else if (store is null)
+            {
+                _count++;
+                entries = new HeaderEntry[InitialCapacity];
+                _headerStore = entries;
+                ref HeaderEntry firstEntry = ref MemoryMarshal.GetArrayDataReference(entries);
+                firstEntry.Key = key;
+                return ref firstEntry.Value!;
+            }
+            else
+            {
+                return ref DictionaryGetValueRefOrAddDefault(key);
+            }
+
+            ref object? GrowEntriesAndAddDefault(HeaderDescriptor key)
+            {
+                var entries = (HeaderEntry[])_headerStore!;
+                if (entries.Length == ArrayThreshold)
+                {
+                    return ref ConvertToDictionaryAndAddDefault(key);
+                }
+                else
+                {
+                    Array.Resize(ref entries, entries.Length << 1);
+                    _headerStore = entries;
+                    ref HeaderEntry firstNewEntry = ref entries[entries.Length >> 1];
+                    firstNewEntry.Key = key;
+                    return ref firstNewEntry.Value!;
+                }
+            }
+
+            ref object? ConvertToDictionaryAndAddDefault(HeaderDescriptor key)
+            {
+                var entries = (HeaderEntry[])_headerStore!;
+                var dictionary = new Dictionary<HeaderDescriptor, object>(ArrayThreshold);
+                _headerStore = dictionary;
+                foreach (HeaderEntry entry in entries)
+                {
+                    dictionary.Add(entry.Key, entry.Value);
+                }
+                Debug.Assert(dictionary.Count == _count - 1);
+                return ref CollectionsMarshal.GetValueRefOrAddDefault(dictionary, key, out s_dictionaryGetValueRefOrAddDefaultExistsDummy);
+            }
+
+            ref object? DictionaryGetValueRefOrAddDefault(HeaderDescriptor key)
+            {
+                var dictionary = (Dictionary<HeaderDescriptor, object>)_headerStore!;
+                ref object? value = ref CollectionsMarshal.GetValueRefOrAddDefault(dictionary, key, out s_dictionaryGetValueRefOrAddDefaultExistsDummy);
```

Comment:
Why can't we just use _ here?

Reply:
Some sort of escape analysis is preventing that:

> CS8347: Cannot use a result of 'CollectionsMarshal.GetValueRefOrAddDefault<HeaderDescriptor, object>(Dictionary<HeaderDescriptor, object>, HeaderDescriptor, out bool)' in this context because it may expose variables referenced by parameter 'exists' outside of their declaration scope

Using a dummy field for it was the only way I found to workaround it.

Reply:
I have no idea what that error means.

@stephentoub Any insight here?

Reply:
You get that error with `out _` but not with `out s_dictionaryGetValueRefOrAddDefaultExistsDummy`?

Reply:
That's right

Reply:
@jaredpar, is this a compiler bug?
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGAzAATakDCpA3oaQ6QA5QCWAbgIYAuM5OSKqQDOAfTABueoxYcefDANLAIEADYjRwKQUalpDEuRSkAsgAoAlAdo29sAGZCAMldIBeAHylHpAHLmAHYwAO5UVgA0pBAArtyawJakOnoAvjY2ioK+1AHUYFGx8Srqyta6jHQVegzAHqTcUDEwKTWGAOw+ME5ikjap+gSpQA==
This compiles, but replace the `out bool s_b` with a discard `out _` and it fails. I'd expect a discard to accommodate any required scope.

Reply:
It's a known issue but up until this point seemed more theoretical than a practical problem. Can look into fixing this with the rest of the `ref` work we're doing this time around.

https://github.com/dotnet/roslyn/issues/56587#issuecomment-934955526

Note: the discussion on that last issue details the trade offs in fixing this. Effectively to fix this we need to ensure that discards have unique locals associated with them when discards mix with span safety rules (today we re-use temps). Otherwise it can lead to discards creating safety issues.